cupy/cuda/memory.pyx

```diff
@@ -243,6 +263,89 @@ cdef class ManagedMemory(BaseMemory):
             runtime.free(self.ptr)
 
 
+@cython.no_gc
+cdef class SystemMemory(BaseMemory):
+    """Memory allocation on an HMM/ATS enabled system.
+
+    HMM stands for heterogeneous memory management. It is a kernel-level
+    feature allowing memory allocated via the system ``malloc`` to be
+    accessible by both CPU and GPU.
+
+    ATS stands for Address Translation Services. It is a hardware/software
+    feature on Grace Hopper that enables the CPU and GPU to share a single
+    per-process page table, allowing memory allocated by the system to be
+    accessible by both CPU and GPU.
+
+    This class provides an RAII interface of the memory allocation.
+
+    Args:
+        size (int): Size of the memory allocation in bytes.
+    """
+
+    def __init__(self, size_t size):
+        self.size = size
+        # TODO(leofang): using the GPU id may not be ideal, but setting it
+        # to cudaCpuDeviceId (-1) would require a lot of changes
+        self.device_id = device.get_device_id()
+        self.ptr = 0
+        if size > 0:
+            self.ptr = <intptr_t>c_malloc(size)
```

Comment:
Q: Is it ok even if `size` is unaligned to 512?

Reply:
Excellent question! We are protected by `_round_size()` which always rounds up to units of 512B before we arrive here.